dot/sync/chain_processor_integration_test.go

```diff
@@ -13,33 +13,127 @@ import (
 
 	"github.com/stretchr/testify/require"
 )
 
-// TODO: add test against latest gssmr runtime
-// See https://github.com/ChainSafe/gossamer/issues/2703
-func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
-	syncer := newTestSyncer(t)
-	responder := newTestSyncer(t)
+func buildBlockWithSlotAndTimestamp(t *testing.T, instance state.Runtime,
+	parent *types.Header, currentSlot, timestamp uint64) *types.Block {
+	t.Helper()
 
-	// get responder to build valid chain
-	parent, err := responder.blockState.(*state.BlockState).BestBlockHeader()
+	digest := types.NewDigest()
+	prd, err := types.NewBabeSecondaryPlainPreDigest(0, currentSlot).ToPreRuntimeDigest()
+	require.NoError(t, err)
+	err = digest.Add(*prd)
 	require.NoError(t, err)
+	header := &types.Header{
+		ParentHash:     parent.Hash(),
+		StateRoot:      common.Hash{},
+		ExtrinsicsRoot: common.Hash{},
+		Number:         parent.Number + 1,
+		Digest:         digest,
+	}
 
-	bestBlockHash := responder.blockState.(*state.BlockState).BestBlockHash()
-	rt, err := responder.blockState.GetRuntime(bestBlockHash)
+	err = instance.InitializeBlock(header)
+	require.NoError(t, err)
+
+	inherentData := types.NewInherentData()
+	err = inherentData.SetInherent(types.Timstap0, timestamp)
+	require.NoError(t, err)
+
+	err = inherentData.SetInherent(types.Babeslot, currentSlot)
+	require.NoError(t, err)
+
+	parachainInherent := babe.ParachainInherentData{
+		ParentHeader: *parent,
+	}
+
+	err = inherentData.SetInherent(types.Parachn0, parachainInherent)
+	require.NoError(t, err)
+
+	err = inherentData.SetInherent(types.Newheads, []byte{0})
 	require.NoError(t, err)
 
-	for i := 0; i < maxResponseSize*2; i++ {
-		block := BuildBlock(t, rt, parent, nil)
-		err = responder.blockState.(*state.BlockState).AddBlock(block)
+	ienc, err := inherentData.Encode()
+	require.NoError(t, err)
+
+	// Call BlockBuilder_inherent_extrinsics which returns the inherents as encoded extrinsics
+	encodedInherentExtrinsics, err := instance.InherentExtrinsics(ienc)
+	require.NoError(t, err)
+
+	// decode inherent extrinsics
+	var inherentExtrinsics [][]byte
+	err = scale.Unmarshal(encodedInherentExtrinsics, &inherentExtrinsics)
+	require.NoError(t, err)
+
+	// apply each inherent extrinsic
+	for _, inherent := range inherentExtrinsics {
+		encodedInherent, err := scale.Marshal(inherent)
+		require.NoError(t, err)
+
+		applyExtrinsicResult, err := instance.ApplyExtrinsic(encodedInherent)
+		require.NoError(t, err)
+		require.Equal(t, applyExtrinsicResult, []byte{0, 0})
+	}
+
+	finalisedHeader, err := instance.FinalizeBlock()
+	require.NoError(t, err)
+
+	body := types.Body(types.BytesArrayToExtrinsics(inherentExtrinsics))
+
+	finalisedHeader.Number = header.Number
+	finalisedHeader.Hash()
+
+	return &types.Block{
+		Header: *finalisedHeader,
+		Body:   body,
+	}
+}
+
+func buildAndAddBlocksToState(t *testing.T, runtime state.Runtime, blockState *state.BlockState, amount int) {
```

Comment:
supernit make `amount` an `uint` (aka restrict dev error as much as possible 😄)

```suggestion
func buildAndAddBlocksToState(t *testing.T, runtime state.Runtime, blockState *state.BlockState, amount uint) {
```